Upgrades: Introduce visuals for wordads instant activation

artpi opened the pull request:
This provides visuals for WordAds-Instant Activation.
It will show "WordAds" element in sidebar for business users and let them activate WordAds instantly

Relevant sections are blocked, only outline is showed.
Copy probably could use some tweaks

# Testing

- Check out this branch
- Connect your sandbox
- Apply Phabricator DIFF D1761
- Create a new premium/business site
- You should have WordAds in Sidebar now
- After togling "Activate" in WordAds settings you should receive an email and now you thould have WordAds enabled. 

##### TODO

- [x] Connect toggle to endpoint
- [x] Reload site in sitesList once WordAds is activated for a site

## Earnings

![zrzut ekranu 2016-05-11 o 09 55 20](https://cloud.githubusercontent.com/assets/3775068/15174096/2307fe54-1760-11e6-9d7f-7bf9271a4cff.png)

## Settings

![zrzut ekranu 2016-05-11 o 09 55 27](https://cloud.githubusercontent.com/assets/3775068/15174102/2b902af6-1760-11e6-8ccf-4dd9086deff6.png)


CC @rralian @mtias @adambbecker @egill @scruffian @dbspringer @drw158

dbspringer commented:
I think that solution looks pretty good (to my not-designer eye) if we're looking to add on to the current design.

My other thought is maybe instead of including the two tabs with the button card on top, what if we added a different page entirely with a more focused call to action? A landing page of sorts with a description of WordAds benefits an activate button?

<img width="632" alt="screen shot 2016-05-09 at 11 16 10 am" src="https://cloud.githubusercontent.com/assets/273708/15123208/8d1bfad2-15d7-11e6-889b-7c57cb386b62.png">

artpi commented:
Implemented copy from editorial team, screenshots updated.

> what if we added a different page entirely with a more focused call to action?

I see that as a part of V1.5 and I think we should do it, but we need some proper design for it, so lets tackle it once @adambbecker or @drw158 finds time for it :)

drw158 commented:
I like @dbspringer 's idea about making it more promotional, but in the meantime, I think a small improvement could be made to make it more consistent with a "settings-style" page. Here are some quick mockups:

![artboard](https://cloud.githubusercontent.com/assets/618551/15221687/1d4fb966-1832-11e6-979b-345911a18baf.png)

![artboard](https://cloud.githubusercontent.com/assets/618551/15221706/2772da2c-1832-11e6-8192-224268b19777.png)

artpi commented:
Updated to @drw158 design:

![screen shot 2016-05-16 at 18 41 25 pm](https://cloud.githubusercontent.com/assets/3775068/15296615/651bc0d4-1b96-11e6-9f21-93b8a34d50ba.png)

retrofox commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -1938,6 +1938,17 @@ Undocumented.prototype.unregisterDevice = function( deviceId, fn ) {
+ * @param {int}       siteId            The site ID
+ * @returns {XMLHttpRequest}          The XHR instance
+ */
+Undocumented.prototype.wordAdsApprove = function( siteId ) {
```
Let's try to add this method directly to wpcom.js. I can do it if you are agree.

artpi replied:
Thanks @retrofox ! For the test, lets leave it in undocumented and if we decide to have it in wpcom, then I will ask you to move it :)

gwwar commented on client/components/plans/plans-compare/index.jsx:
```diff
@@ -233,6 +234,14 @@ const PlansCompare = React.createClass( {
 			const plans = this.getPlans();
 			const features = this.getFeatures();
 
+			abtest( 'wordadsInstantActivation' ) === 'enabled' && features.splice( 6, 0, {
```
Should we also check for `config.isEnabled( 'manage/ads/wordads-instant' )`? Folks might be able to see copy, but won't be able to activate if they upgrade.

artpi replied:
done!

gwwar commented on client/state/wordads/approve/reducer.js:
```diff
@@ -0,0 +1,52 @@
+} from 'state/action-types';
+
+/**
+ * Tracks all known user objects, indexed by user ID.
```
Copy pasta?

gwwar commented on client/state/wordads/approve/reducer.js:
```diff
@@ -0,0 +1,52 @@
+	error: null
+};
+
+export function requestStatus( state = initialState, action ) {
```
Are word ad requests per user, or per site? If it's the latter I'd suggest indexing by site id, or clearing state when switching sites.

rralian replied:
They are per site.

gwwar commented:
Are there testing instructions available? I'd be happy to help test this after a rebase.

gwwar commented on client/state/sites/reducer.js:
```diff
@@ -40,6 +41,14 @@ const VALID_SITE_KEYS = Object.keys( sitesSchema.patternProperties[ '^\\d+$' ].p
  */
 export function items( state = {}, action ) {
 	switch ( action.type ) {
+		case WORDADS_SITE_APPROVE_REQUEST_SUCCESS:
```
Why are we storing state on the site subtree, when we've also added a root `state/wordads` tree?

artpi replied:
Yes, we updated the `options.wordads` property of `/me/sites`.

I kept it flat, in `state/wordads`, because everything else is like it - it's even "glued together" in sidebar.
I personally hate this and I think we should make better integration.
But folks are of the opinion that for now it should be separate, so I'm following that pattern here.
After the test, if we decide to keep it, I'm happy to integrate better.

gwwar replied:
It's just a little odd since the interesting state information is being set on sites, rather than in this tree. I think it's ok to leave this here if we think we'll be storing more information here in the future, otherwise we can have a follow up PR to move this to a more appropriate location.

artpi replied:
We cannot move this particular piece of information anywhere, because this is supposed to be here in the sites-list endpoint.
It was here all along.
I'm just setting it here manually, because I dont want to reload the sites endpoint, but `options.wordads` is an information that was always in API response.

gwwar replied:
I was referring to the `state/wordads` tree. The option data is fine here.

lamosty commented:
I'm trying to test this too and after checking the branch and enabling the `wordadsInstantActivation` AB test, the WordAds menu item is not in the main navigation menu (sidebar) on the left. How can I test it? :)

artpi commented:
> How can I test it? :)

Good question! I added "Testing section" to the top post.
The way it works right now is that we need to merge Phabricator diff `D1761` to wpcom.
That diff introduces `activate_wordads` capability to `me/sites` endpoint. Wordads shows in sidebar only when you have this capability.

Alternatively, you can comment out this line https://github.com/Automattic/wp-calypso/pull/5292/files#diff-725f2c27250eb900e666e4ad67d51305R34 - then you will have Wordads in the sidebar for any premium site you own, but activation wont work.

lamosty commented:
> Good question! I added "Testing section" to the top post.

Thanks!

Working for me :+1:.

![selection_004](https://cloud.githubusercontent.com/assets/4988512/15705292/e4827d4a-27ee-11e6-97a8-b6cd1216b5aa.png)

![selection_005](https://cloud.githubusercontent.com/assets/4988512/15705294/e7db2474-27ee-11e6-91d4-70bbc69b1907.png)

artpi commented:
@lamosty, thanks for testing!

I tweaked the email in the phabricator diff:
<img width="738" alt="zrzut ekranu 2016-06-01 o 13 42 40" src="https://cloud.githubusercontent.com/assets/3775068/15714047/44142aae-2819-11e6-940f-35ebc2c0b800.png">

gwwar commented on client/my-sites/ads/main.jsx:
```diff
@@ -60,16 +72,69 @@ module.exports = React.createClass( {
+				}
+				{ this.props.requestingWordAdsApproval &&
+					<Notice status="is-info ads__activate-notice" showDismiss={ false }>
+						{ this.translate( 'Kindly requesting WordAds activation' ) }
```
Maybe 'Requesting WordAds activation' here? Editorial might have some better ideas

artpi replied:
I disagree withe the oppresive urge to strip any emotion! :)

artpi commented:
After speaking with @rralian we decided to move entry point to this flow from Sidebar to My-Plan page:
<img width="806" alt="zrzut ekranu 2016-06-02 o 17 13 01" src="https://cloud.githubusercontent.com/assets/3775068/15750590/601a4c0a-28e7-11e6-923a-1b2ee3d716d3.png">

apeatling commented:
Feedback from P2:

- Scrolling is a big issue, clicking on buttons on the my plan page, and the “missing features” link at the bottom does not scroll the person to the top on the next page, that is really disorienting.
- The toggle is not the right element to use here. You are never going to be able to toggle this off so it doesn’t make a lot of sense. It should just be a button “Enable Wordads”.
- We try and be optimistic in Calypso. I don’t see any reason why we can’t just show Wordads turned on for the user, rather than showing the “requesting” message. The toggle should also switch to “on” straight away. If there was an error we can show that and disable it again. That would mean 99.9% of people don’t have to wait.
- I think there needs to be a better transition between turning this on and then seeing the settings pages become active. It is very jarring seeing the box disappear. There should be a message saying that it is now enabled, and maybe we can slide the box away, or do something a little less jumpy.

gwwar commented:
Re: the scroll issue, it looks like we manually fix this for sidebar items with a [window.scrollTo](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/sidebar/sidebar.jsx#L53). @mtias I think we'll want a general solution for Calypso links, but I'll add in a fix for the my-plans page first.

artpi commented:
New video with the flow: https://videopress.com/v/Uclws4aC

- Added new flow with success messaging
- Tweaked redux tree
- Moved from toggle to button

gwwar commented on client/state/wordads/approve/reducer.js:
```diff
@@ -0,0 +1,94 @@
+				[ action.siteId ]: action.error
+			} );
+		case WORDADS_SITE_APPROVE_REQUEST_SUCCESS:
+		case WORDADS_SITE_APPROVE_REQUEST_DISMISS_ERROR:
```
We forgot to reset error state, when another `WORDADS_SITE_APPROVE_REQUEST` is triggered.

artpi replied:
Will fix, thanks!

gwwar commented on client/state/wordads/approve/actions.js:
```diff
@@ -0,0 +1,45 @@
+};
+
+export const dismissWordAdsError = ( siteId ) => ( { type: WORDADS_SITE_APPROVE_REQUEST_DISMISS_ERROR, siteId } );
+export const dismissWordAdsSuccess = ( siteId ) => ( { type: WORDADS_SITE_APPROVE_REQUEST_DISMISS_SUCCESS, siteId } );
```
This works but we could also just reset request state, when `WORDADS_SITE_APPROVE_REQUEST` is dispatched.

gwwar replied:
An exception to that would be if we need the error state to be displayed to the user while we trigger another wordads request

artpi replied:
Actually not - I dismiss this when user updates settings which is in a completely different place

gwwar replied:
Gotcha 👍

gwwar commented on client/state/wordads/approve/reducer.js:
```diff
@@ -0,0 +1,94 @@
+export function requestSuccess( state = {}, action ) {
+	switch ( action.type ) {
+		case WORDADS_SITE_APPROVE_REQUEST_DISMISS_SUCCESS:
+		case WORDADS_SITE_APPROVE_REQUEST_FAILURE:
```
We can also reset state here on  `WORDADS_SITE_APPROVE_REQUEST`

gwwar commented on client/state/wordads/approve/reducer.js:
```diff
@@ -0,0 +1,94 @@
+
+export default combineReducers( {
+	requesting,
+	requestSuccess,
```
Were you having issues with using the requesting and error state? e.g. `const success = ! this.props.isRequesting && ! this.props.wordAdsError`. This feels a little redundant, unless we need extra state for the success error message ( for example we get back different success messages ). This works, but just be careful to not get into a conflicting state with requestErrors.

artpi replied:
I am displaying success message when success has gone through with information what to do next.
If I use `this.props.isRequesting && ! this.props.wordAdsError`, then when you activated long time ago, this condition is still true.
I want display success only when you have "just activated"

gwwar commented on client/state/wordads/approve/selectors.js:
```diff
@@ -0,0 +1,68 @@
+}
+
+/**
+ * Returns true if we're currently requesting WordAds approval
```
```Returns true if the WordAds approval request was successful```

artpi replied:
Will fix, thanks!

gwwar commented:
Thanks for adding test for the new redux bits 😄

rralian commented:
I made a handful of tweaks. I think once you address the issues @gwwar brought up this is good to merge.

artpi commented:
>  once you address the issues @gwwar brought up this is good to merge.

Addressed. (also few other tweaks)
Also, I enabled for wpcalypso on this PR so it will be easier to reason about.

@rralian : I have one concern about you changing the test here:
Presumably we can significantly increase number of sites in wordads with this PR - thats why I enabled it only for 5% at the beginning to monitor any problems there and ramp it up.
But, lets merge this PR and start this discussion in the PR enabling it in prod.

rralian commented:
> Presumably we can significantly increase number of sites in wordads with this PR

If we're running the test to reach significance, we're not getting any more money by running at 95/5. It will just take us ten-times as long to get results... we need to reach the minimum sample size for each variant regardless of the percentage split in the test. So ultimately the same number of people will see this test as "off" (and are therefore a missed upsell opportunity) regardless of the split we run. By doing 50/50 however we can run the test for about two weeks and be done with it (versus two and a half months).

rralian commented:
I think we're ready to merge and get this viewable by other a11ns in wpcalypso.

rralian commented:
I rebased and will do a squash-merge now.
